src/cipher.c

```diff
@@ -641,20 +612,149 @@ static int p11prov_aes_get_ctx_params(void *ctx, OSSL_PARAM params[])
+                return RET_OSSL_ERR;
+            }
+            /* Currently only CS1 is supported */
+            if (strcmp(mode, OSSL_CIPHER_CTS_MODE_CS1) != 0) {
+                CK_RV rv = CKR_MECHANISM_PARAM_INVALID;
+                P11PROV_raise(ctx->provctx, rv, "Unsupported mode: %s", mode);
+                return RET_OSSL_ERR;
+            }
+        }
+    }
+
+    return RET_OSSL_OK;
 }
 
-static const OSSL_PARAM *p11prov_aes_gettable_ctx_params(void *ctx,
+static const OSSL_PARAM p11prov_aes_generic_gettable_ctx_params[] = {
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_KEYLEN, NULL),
+    OSSL_PARAM_size_t(OSSL_CIPHER_PARAM_IVLEN, NULL),
+    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_PADDING, NULL),
+    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_NUM, NULL),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_IV, NULL, 0),
+    OSSL_PARAM_octet_string(OSSL_CIPHER_PARAM_UPDATED_IV, NULL, 0),
+    /* Supported by OpenSSL but not here yet
+     * OSSL_CIPHER_PARAM_TLS_MAC
+     */
+    OSSL_PARAM_END
+};
+
+static const OSSL_PARAM *p11prov_aes_gettable_ctx_params(void *vctx,
                                                          void *provctx)
 {
+    struct p11prov_cipher_ctx *ctx = (struct p11prov_cipher_ctx *)vctx;
+
+    if (!ctx) {
+        return p11prov_aes_generic_gettable_ctx_params;
+    }
+
+    switch (ctx->mech.mechanism) {
+    case CKM_AES_ECB:
+    case CKM_AES_CBC_PAD:
+    case CKM_AES_OFB:
+    case CKM_AES_CFB128:
+    case CKM_AES_CFB1:
+    case CKM_AES_CFB8:
+    case CKM_AES_CTR:
+    case CKM_AES_CTS:
+        return p11prov_aes_generic_gettable_ctx_params;
+    }
     return NULL;
 }
 
-static const OSSL_PARAM *p11prov_aes_settable_ctx_params(void *ctx,
+#define GENERIC_SETTABLE_CTX_PARAMS() \
+    OSSL_PARAM_uint(OSSL_CIPHER_PARAM_PADDING, NULL)
+/* Supported by OpenSSL but not here:
+ * OSSL_CIPHER_PARAM_NUM (uint)
+ * OSSL_CIPHER_PARAM_USE_BITS (uint)
+ * OSSL_CIPHER_PARAM_TLS_VERSION (uint)
+ * OSSL_CIPHER_PARAM_TLS_MAC_SIZE (size_t)
+ */
+
+static const OSSL_PARAM p11prov_aes_generic_settable_ctx_params[] = {
+    GENERIC_SETTABLE_CTX_PARAMS(), OSSL_PARAM_END
+};
+
+static const OSSL_PARAM p11prov_aes_cts_settable_ctx_params[] = {
+    GENERIC_SETTABLE_CTX_PARAMS(),
+    OSSL_PARAM_utf8_string(OSSL_CIPHER_PARAM_CTS_MODE, NULL, 0), OSSL_PARAM_END
+};
+
+static const OSSL_PARAM *p11prov_aes_settable_ctx_params(void *vctx,
                                                          void *provctx)
 {
+    struct p11prov_cipher_ctx *ctx = (struct p11prov_cipher_ctx *)vctx;
+    if (!ctx) {
+        return p11prov_aes_generic_settable_ctx_params;
+    }
+    switch (ctx->mech.mechanism) {
+    case CKM_AES_ECB:
+    case CKM_AES_CBC_PAD:
+    case CKM_AES_OFB:
+    case CKM_AES_CFB128:
+    case CKM_AES_CFB1:
+    case CKM_AES_CFB8:
+    case CKM_AES_CTR:
+        return p11prov_aes_generic_settable_ctx_params;
+    case CKM_AES_CTS:
+        return p11prov_aes_cts_settable_ctx_params;
```

Comment:
I see in the skey test only the AES-CBC. Should we add tests for the other modes we support (might need some allow/block-listing for pkcs11 backends though.

Reply:
We still have a long way to go with skey support, could you file a ticket for this?

Reply:
Sure. Filled #586.